pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ProducerBuilder.java

```diff
@@ -608,4 +608,15 @@
      * @return the producer builder instance
      */
     ProducerBuilder<T> enableLazyStartPartitionedProducers(boolean lazyStartPartitionedProducers);
+
+    /**
+     * This config will ensure that If possible PartitionedProducer would attempt to produce message on
+     * another available partitions, If currently picked partition is not available for some reason.
+     * Next available partition will be chosen by the same routing policy as client is configured with.
+     * @param maxRetryOtherPartition
+     *            How many partitions should be tried before bailing out
+     * @return the producer builder instance
+     */
+    ProducerBuilder<T> maxRetryOtherPartitions(int maxRetryOtherPartition);
```

Comment:
Instead of having a configuration here, we should probably extend the router interface, since the possibility of re-routing is very tied to a particular router implementation

Reply:
Actually. Initially thought about this and decided against it for two reasons.
1. Main criteria for rerouting is whether a partition is available or not. I thought leaking that information into routing side  was not neat. So much so that even I did not like leaking even batching part into Routing.
2. Also rerouting or not is based on transaction etc so I guess I would not be comfortable leaking that information in routing. 

Algorithm is: Hey, this partition is not available. Give me another partition to publish. If next partition is same as the current one. Bad.. But if next partition is different. Good. Let's publish to the partition.

Reply:
I agree with @merlimat 

The issue is the RoundRobin router. Not for all the routers.
IMO, it looks like we need a new router based on the available pending messages.

e.g. we have 3 partitions

0 -> 1000 available pending messages
1 -> 1000 available pending messages
2 -> 1000 available pending messages

After sending 3 messages.

0 -> 999 available pending messages
1 -> 999 available pending messages
2 -> 999 available pending messages

After receiving the send receipt 

0 -> 1000 available pending messages
1 -> 999 available pending messages (partition-1 is unavailable)
2 -> 1000 available pending messages

The subsequent new messages should choose partition-0 and partition-1 and so on.

Reply:
@codelipenghui I am not sure I understood. But I have not touched this behavior. Routing behavior if Round Robin router is not changed. Only thing that is changing is. When new message is choosing partition-0 and partition-0 is not READY. New message should ask router to give another partition and another until either of the three situiation happen.

- Router returns a partition which is connected.
      OR
- Router finds a partition and its same on, in case of transaction/keyed/single partition so no retry  
     OR
- Or returns a partition which is not connected. Retry and ask router to give another partition, until configurable number of time.